server/src/main/java/org/elasticsearch/synonyms/SynonymsManagementAPIService.java

```diff
@@ -102,7 +109,52 @@ private static XContentBuilder mappings() {
         }
     }
 
-    public void putSynonymsSet(String resourceName, SynonymsSet synonymsset, ActionListener<PutSynonymsAction.Response> listener) {
+    public void getSynonymsSet(String resourceName, int from, int size, ActionListener<SynonymsSetResult> listener) {
+        client.prepareSearch(SYNONYMS_ALIAS_NAME)
+            .setQuery(QueryBuilders.termQuery(SYNONYMS_SET_FIELD, resourceName))
+            .setFrom(from)
+            .setSize(size)
+            .execute(listener.delegateFailure((searchResponseListener, searchResponse) -> {
+                final long totalSynonymRules = searchResponse.getHits().getTotalHits().value;
+                if (totalSynonymRules == 0) {
+                    listener.onFailure(new ResourceNotFoundException("Synonym set [" + resourceName + "] not found"));
```

Comment:
This is indeed a problem, as we can't tell apart an empty synonym set from one that does not exist in the current design.

I think it's important to return a 404 for resources that do not exist; it would be confusing for the users otherwise.

We should avoid empty result sets:
- Reject a PUT request without synonym rules
- Remove automatically a synonym set when the last rule is removed

In case we need to support empty synonym sets, I think we would need to provide separate documents to represent synonym sets in the system index. I would prefer to avoid that, at least until we have a working solution.

What are your thoughts?

Reply:
What action does the [synonym token filter](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html) take if the synonym file is empty? Should there be parity with whatever is currently supported?

FWIW I'm fine with not allowing empty synonym sets since it seems pointless, but this could also be interesting in a future UI - you can never remove the last synonym from a set.

Reply:
Good! Then it should be taken care of already, as the PUT Request validation already includes checking that the synonym set is not empty.

We can remove the synonym set when the last rule is deleted when we implement the DELETE for individual rules.

Reply:
@kderusso I've checked what what is happening with synonym token filter with files:
- If a file is absent, an error is raised
- If a file is present and empty, there is no error,  and a synonym filter is initialized successfully with an empty list of synonyms.

Reply:
@carlosdelest  

> We can remove the synonym set when the last rule is deleted when we implement the DELETE for individual rules.

Can you please explain what do you mean by "we can remove the synonym set"? As we don't have synonym set outside of documents for individual rules.

Reply:
Yes, I meant it won't exist anymore.